darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -28,6 +29,53 @@
 from darshan.experimental.plots import heatmap_handling
 
 
+def determine_hmap_runtime(report: darshan.DarshanReport) -> Tuple[float, float]:
+    """
+    Determine the effective heatmap runtime to be used
+    for plotting in cases where only DXT, only HEATMAP,
+    or both module types are available, to achieve a common
+    max displayed runtime.
+
+    In some cases, this may mean that the time value is
+    rounded up from the actual runtime.
+
+    Paramaters
+    ----------
+
+    report: a ``darshan.DarshanReport``
+
+    Returns
+    -------
+
+    A tuple containing `tmax`, (rounded) `runtime` floats.
+
+    """
+    # calculate the elapsed runtime
+    runtime = report.metadata["job"]["end_time"] - report.metadata["job"]["start_time"]
+    # ensure a minimum runtime value of 1 second
+    runtime = max(runtime, 1)
+    # leverage higher resolution DXT timing
+    # data if available
+    if ("DXT_POSIX" in report.modules or
+        "DXT_MPIIO" in report.modules):
+        tmax = 0.0
+        for mod in report.modules:
+            if "DXT" in mod:
+                agg_df = heatmap_handling.get_aggregate_data(report=report,
+                                                             mod=mod,
+                                                             ops=["read", "write"])
+                tmax_dxt = float(agg_df["end_time"].max())
```

Comment:
This branch now vastly outperforms `pydarshan-devel` on this "benchmark" despite having to do substantially more work:

- `pydarshan-devel`: `2:08.91`
- this branch: `1:15.89`

This goes back to my [original comment](https://github.com/darshan-hpc/darshan/pull/396#issuecomment-882115047) almost a year ago that the DXT processing algorithm was merely a first draft that merited careful scrutiny/refinement.

As far as why CI will now pass all versions of Python other than 3.6, haven't checked that, but can always add the old code path back for that Python version, and still be far more performant on the Python versions that are not end-of-life while doing more work here.

Reply:
As a side note on selecting the types of benchmarks you review with, careful what you wish for, because on this branch this will probably select pretty strongly for that single large case...

`asv continuous -e -b ".*heatmap.*" pydarshan-devel treddy_issue_652_runtime_hmap_report`

```
       before           after         ratio
     [543191e6]       [b4c59a2c]
     <pydarshan-devel>       <treddy_issue_652_runtime_hmap_report>
+      51.4±0.7ms          285±8ms     5.54  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 250, 1.0)
+        63.4±5ms         286±10ms     4.51  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 250, 0.01)
+        49.0±2ms          188±3ms     3.84  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 250, 0.001)
+        51.0±4ms          191±2ms     3.74  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 250, 0.001)
+         599±4ms       2.23±0.08s     3.73  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 50, 1.0)
+        533±20ms       1.93±0.03s     3.63  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 10, 1.0)
+        62.5±2ms          224±4ms     3.58  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 10, 1.0)
+        77.4±4ms          276±7ms     3.56  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 50, 1.0)
+        54.3±2ms          189±5ms     3.49  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 250, 0.01)
+        54.6±3ms         188±10ms     3.44  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 250, 0.001)
+      23.8±0.4ms       80.7±0.6ms     3.39  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 50, 0.01)
+        20.6±1ms         69.7±1ms     3.38  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 50, 1.0)
+        55.4±3ms          186±1ms     3.35  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 250, 0.01)
+         160±7ms         519±10ms     3.25  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 250, 1.0)
+      17.1±0.3ms         48.7±2ms     2.84  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 50, 0.001)
+        19.1±1ms         51.6±2ms     2.70  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 50, 0.001)
+      1.08±0.05s          2.90±0s     2.69  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 250, 1.0)
+      17.7±0.5ms         47.1±1ms     2.66  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 50, 0.01)
+        18.9±1ms         49.9±2ms     2.64  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 50, 0.01)
+        19.1±2ms         49.8±1ms     2.60  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 50, 0.001)
+        14.4±1ms         31.8±1ms     2.20  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 10, 1.0)
+        18.4±1ms         40.0±2ms     2.17  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 10, 0.01)
+      11.4±0.8ms         21.2±2ms     1.85  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 10, 0.01)
+      10.8±0.3ms       19.6±0.9ms     1.82  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 10, 0.001)
+      11.1±0.4ms         19.3±2ms     1.74  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(1000, 10, 0.001)
+      11.0±0.1ms       18.6±0.9ms     1.70  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(50, 10, 0.01)
+      11.7±0.7ms       19.8±0.2ms     1.69  dxt_heatmap.GetHeatMapDf.time_get_heatmap_df(10000, 10, 0.001)
+            387M             425M     1.10  dxt_heatmap.GetHeatMapDf.peakmem_get_heatmap_df(10000, 250, 1.0)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```